### ISSUE TITLE
Ad-Blocker VPN mode stops working when switching connection type

### ISSUE THREAD

**Issue by elvissteinjr** [AdAway/AdAway#2154]
**Describe the bug**
Switching from wifi to mobile internet or vice versa by turning off the other connection type makes the VPN not load anything anymore. Simply pausing and unpausing from the notification to restart the VPN quickly solves it, but it would be nice if the connection change could possible detected and have this happen automatically.
It doesn't break entirely rather than just not switching networks it seems, as switching to back to the previously working connection works without restarting the VPN in-between.

**To Reproduce**
Using AdAway in VPN mode:
- Start VPN
- Turn off Wifi and enable mobile data if it's not on yet
- Try to load anything from the web
- Restart the VPN to fix the issue manually

**Expected behavior**
The VPN should switch connections automatically if possible.

**Device**
 - AdAway version: 5.0.6
 - Device: F(x)tec Pro1
 - Android version: 9
 - Root version: Unrooted

**Additional context**
This seems like a weird thing to miss, so I'm wondering if there's some kind of limitation about implementing this or if there's some kind of bug in the OS of my phone (wouldn't be surprised, tbh). 
I can provide logs if there's something I should be looking out for. ConnectivyService and the likes seem to log the connection change fine at least, but otherwise there's nothing standing out to me (and nothing from AdAway for that matter).

I'd also like to say: When VPN mode is working, it really works just fine. Like, multiple days without issues. As long as I don't need to access anything a different connection as described. So nice work there!

**Comment by PerfectSlayer** [comment]
Hi @elvissteinjr 

Thanks for the report. I was asking some feedback about VPN ad blocking method lately so it's perfect.
In fact, I am only testing the VPN method on my test device which does not have a sim card. So I do never disable wifi otherwise I do no more have internet access.

It's the kind of the bug I want to fix quickly before moving the beta to stable 😉

**Comment by PerfectSlayer** [comment]
I tried to reproduce on my device and I can't trigger the issue.
I tried from mobile network to wifi, wifi to mobile network, mobile network to airplane mode and airplane mode to mobile network.

If you can provide me some logcat, I think it would be helpful 👍 
In the mean time, I will keep using as a daily ad-blocker and see if it happens on my device.

**Comment by elvissteinjr** [comment]
I've recorded some [logs](https://github.com/AdAway/AdAway/files/5114845/2020-08-23-21-45-10.txt), but since I don't know exactly what to look for (the few lines that do mention AdAway are not interesting at least), there's a lot of noise in there.
What I did when recording the log was: Turn off wifi -> turn on mobile internet -> try to access google.com (failed) -> restart AdAway VPN from the notification -> try to access google.com (success).

Now what got me stumped is that right after recording the first attempt, which went as expected, I could not reproduce my issue  again. I were suddenly able to clearly observe the notification updating the VPN state right when a connection got changed too.
Yet, after about 10 minutes the previous behavior returned. 
I had been running AdAway behaving as described in the first post for more than a month with no significant changes so it's odd and this was the first time I've observed it working as intended like this.

So is this power-saving related? My phone's OS doesn't do anything special and is pretty much stock Android in that regard to my knowledge. Other apps keeping a notification to stay active don't seem to have any issues though, so I'm just guessing around here.

Typing this up I'm getting the feeling this might be out of your hands and the log may not provide much useful information either. I apologize for wasting your time if that's the case.
Either way, thanks for looking at this.

**Comment by fawkes15** [comment]
I have the same bug on OnePlus 8 Android 10 no root.

**Comment by ipdev99** [comment]
Totally random for me.
I will go days without it happening and then it will happen once or multiple times in one day.
Sometimes adaway appears to be running (red background).
Other times adaway is paused (grey background).
Simple pause and start (or just start) will solve the issue.

Not currently reproducible at will.

Is it related to adaway starting paused on my device?
(AdAway starts on boot but, it starts paused not active.
Reproducible but, the issue might still be on my side.

It has been a few weeks since it last happened.
Today it happened twice. A few hours after I got to work and again when I got home.
Prior logs did not show much. Today's look like they might be some help.

Log_2020-9-26_09-53-08 - Taken with LogcatExtream when I noticed it this morning.
adblogcat - When I got home and noticed it was not connecting again.
bugreport - Accidentally killed it before it was done.

Did not think about running bugreport this morning.
Next time it happens, I will run it from terminal.

Sorry for full logs, since I can not reproduce at will, I can not clear the log before hand.

Cheers.

[adblogcat.log](https://github.com/AdAway/AdAway/files/5287517/adblogcat.log)
[Log_2020-9-26_09-53-08.txt](https://github.com/AdAway/AdAway/files/5287519/Log_2020-9-26_09-53-08.txt)

**Comment by BearTM** [comment]
I'm regularly experiencing the same issue. The VPN appears to not detect or manage the network switch, and remains "active" on the non-existent network.

Pause/unpause, connect/disconnect, flight mode on/off, ... Any action will restore connectivity.

Happens often for me when leaving home. Move out of range of WiFi (switches to mobile ok), drive to destination, reconnect to different WiFi (fails & VPN blocks all traffic).

Pixel 5, non-root, latest updates.

Maybe something as simple as on repeated traffic failure (for DNS lookup) to perform a check of active networks and reconnect?

Otherwise the ad blocking is working very well. Have not noticed any overhead, battery drain or slowdown - only the loss of connectivity on network change.

**Comment by pxcask** [comment]
Hi, I have the same issue on my Fairphone2 with /e/OS. I could not find any workaround, except switching off the VPN and on again. Phone is not routed.

**Comment by abdulocracy** [comment]
This issue has been open for a long time, I'd recommend DNS66 until it's solved.

**Comment by RobBeyer** [comment]
I thought I'd **\*ping\*** this issue again since there's been some more traction with others and error reporting ...

The issue happens ONLY when changing from one network to another. I've had it happen even for Wifi to Wifi, but more often/reliably with Wifi to 4G/5G or the reverse.

For some reason the network change is not detected, and the VPN stays active, routing to the prior network. Turning off & on the VPN resolves it immediately, without any future or further issues. (no crash, no memory issues, no lockup, nothing - runs great without any restart)

Can something "simple?" be implemented which monitors the current active connection?
- Check the active (physical) connection IP address and/or gateway information for any changes?
  - Unfortunate if you change to the same IP address on the new connection, but that's insanely unlikely to happen!
- Attempting a "ping" to an endpoint every minute?
  -  Add config options: time and host to ping? or just ping the connection gateway? Keep-alive?
  - Assuming a 32 byte ping packet every minute, this would cost at most 48k traffic in a day, which I'd happily afford for 99%+ reliability.
- Restart the VPN on repeated DNS lookup failures for valid hosts?
  - On the 3rd DNS lookup timeout failure, attempt a restart of the VPN and/or provide warning/notification of error/issue?

There are probably other possibilities, but anything which can detect failure and simply restart the VPN would be preferred to having the VPN "dead" to traffic, and only noticing the issue a few hours later.

Cheers!

**Comment by PerfectSlayer** [comment]
@abdulocracy Yes, and I still had struggle to find the issue. As you can see, I am alone on the project on my spare time and I still deliver update. I just didn't had the time to find the workaround yet.

Thanks @RobBeyer There is some good ideas here. I will also have a look to what DNS66 does in order to handle this case.
There is so much more I want to do with VNP ad-blocking method (like private DNS client, VPN chaining, etc...) but again, I still lake time in my personal life.

**Comment by PerfectSlayer** [comment]
Let's try something!
I just build a 5.3.0 based version where you can enable the VPN monitor (watchdog).
It is a mechanism that will send an empty DNS request to your resolver to check the network state if there is no network activity.
If the request fails, it will restart the VPN and keep trying to send a DNS request.
If the requests keep failing, it will restart later with a greater delay each time so it won't spam your network and should be battery friendly.

[AdAway-5.3.0-vpnmonitor.zip](https://github.com/AdAway/AdAway/files/5851165/AdAway-5.3.0-vpnmonitor.zip)

_This build is signed with the same key the GitHub releases use. You can install over a GitHub release and it will be updated like any other build when a new update comes._

**Comment by fawkes15** [comment]
I just tried it. I switched several times wifi <> LTE and no issue. 

Thanks a lot!

**Comment by RobBeyer** [comment]
Awesome and a super fast implementation! I will test over the next week and report back if I encounter any issues.

I understand the "time pressure issue" for any side projects, and I greatly appreciate the time you've taken to resolve this issue. I myself still have a hardware project (automatic time-based fish-tank RGB lighting with "weather" and power control for pumps, air, and UV sterilizer) which is 75%+ done, and never manage to "create" the additional time I need to complete it. I just might do that now. :)

Cheers! and thank you.

**Comment by PerfectSlayer** [comment]
You are welcome. I already work on it but never found the time to complete it.
I am really interested to have your feedback in few days to know if I must ship it in a stable build.

Oh, and I also have a small hardware (mainly to keep practicing iron and 3D model/printing): a small Cyberpunk logo  lamp based on an esp88266 and a ws2812 rgb led. I am still on prototyping but spending few hours to iron leds relaxes me 😄

**Comment by RobBeyer** [comment]
Well, I've run through a few tests and while they all worked, I went on a drive to pick up some takeaway food. While the switch to 4G went perfectly, the switch back to my WiFi at home was not detected. The DNS polling did not appear to solve the VPN lockup. All traffic was dead - no apps worked - Chrome, GMail, etc. ... including attempting to update the Blocklists in AdAway.

![Failed to download hosts](https://user-images.githubusercontent.com/45636356/105432527-c72b9680-5c25-11eb-9b32-d02bb56f907a.jpg)
![VPN Configuration](https://user-images.githubusercontent.com/45636356/105432536-cbf04a80-5c25-11eb-931c-2b3c6b464f86.jpg)
![AdAway Preferences](https://user-images.githubusercontent.com/45636356/105432545-cdba0e00-5c25-11eb-99de-0d02bf357409.jpg)

I'm ready to help debug this issue, as it appears I can repeat the issue "somewhat" reliably ... even though I have no idea what I need to do to trigger it.

I've enabled the "Verbose Logging" flag in the app, which I hope might help track down the issue.

I've also installed a (non-root) [Logcat](https://play.google.com/store/apps/details?id=com.tananaev.logcat) and enabled the adb tcpip server on my phone. Since a logcat would be capturing an untold number of messages, can you give me a filter list (tag/keyword) of likely messages that I can use, and leave the log running on my phone until the problem happens.

I hope we can track down the source of this issue.

---

I know exactly what you mean about "practicing iron and 3D model/printing". My current project is the Tardis from Dr. Who - I'm just printing the 4th and last transparent window now.

I have to THANK YOU for the idea of using an ESP8226 and a WS2812 RGB!!! I was just going to use a few white LEDs, but I've got a few ESP8266-01S boards lying around, and having a "WiFi / web-controlled Tardis" just sounds so much cooler!

![Dr. Who - Tardis](https://user-images.githubusercontent.com/45636356/105432603-eb877300-5c25-11eb-843e-289336c42b5c.jpg)

My project just became a little bit bigger ... 👍

**Comment by PerfectSlayer** [comment]
Good to know it improves the overall behavior. Did you wait few minutes after switching back to your home wifi?

About logging, if the logcat app allows you to filter by application, you can filter the `org.adaway` package name.
If not, you can log the following tags: `VpnWatchDog`, `VpnWorker`, `DnsMapper`, `DnsPacketProxy`.

---

You're welcome! And if you want an easy-to-use but advanced control about your led, have a look to [the wled project](https://github.com/Aircoookie/WLED) 😉  Nice setup by the way, I am sort of jealous of your oscilloscope 😅

**Comment by RobBeyer** [comment]
I've had the issue happen again - VPN lockup when returning home and having WiFi re-enable. NOTE: I also have 2 sim cards in my phone - so there are 3 potentially active connections, but a maximum of 2 data connections at any time (Wifi & 4G).

For some reason the adb server was closed on the logcat app (no idea why!?!), so I have no log yet. I'll try again and add in the filtering suggestions and see what is recorded. I probably just ran out of log space leaving it running continuously. 

Thanks for the link to the WLED project ... I will definitely use that for the Tardis. 👍 

For the scope, I have a [Hantek DSO5102P](https://www.aliexpress.com/item/32787528697.html) purchased through AliExpress from the [Hantek Official Store](https://www.aliexpress.com/store/2792111). Excellent store, no issues with purchase or shipping - very well packed. They've got a new [DSO2D15](https://www.aliexpress.com/item/1005001844918818.html) with built-in RS232/UART, I2C, SPI, CAN, LIN decoding with an 8M capture buffer instead of the 40K on mine, which I think is pretty sweet. The review from a German purchaser indicates that the firmware is still being worked on (the future is in beta!), but otherwise it looks like the 'perfect' MCU/embedded hobby-level DSO ... if you're considering one.

**Comment by RobBeyer** [comment]
Something odd to check into ...

VpnWatchDog: handleTimeout is reporting very odd elapsed times:

"Milliseconds elapsed between last receive and sent: 1611366365245"
"Milliseconds elapsed between last receive and sent: 8973"
"Milliseconds elapsed between last receive and sent: -16019"
"Milliseconds elapsed between last receive and sent: -6473645"

Also, a Warning message: handleTimeout(VpnWatchdog.java:130) with nothing more in the log.

I tried to look for the new source, to see if I could spot something, but couldn't find a branch on the GitHub.

There may be an issue with the last receive/sent calculation which is preventing the failure detection?

**Comment by RobBeyer** [comment]
Logcat - Warnings and some exceptions with reconnections. Everything is still working fine, even with the errors in the log. I have not switched networks off WiFi throughout this time.

[logcat.txt](https://github.com/AdAway/AdAway/files/5859561/logcat.txt)

**Comment by PerfectSlayer** [comment]
Thanks for the logcat!
From the log, we can see that the watchdog times out three times (meaning it can't reach network multiple time and tries to restart the VPN service): at 20:32, 20:20:46 and 20:50.

The elapsed time are "_normal_" if you don't receive packet for the request the device send (the watchdog keep alive packets or any internet response packet). It will increase a penalty time the device wait before restarting the VPN.

I push the code on the `master` branch if you would like to have a look. It mainly takes place in the `VpnWatchdog` and `VpnWorker` classes.

**Comment by RobBeyer** [comment]
I still find it odd that the "Milliseconds elapsed between last receive and sent: 1611366365245" can be so HUGE (it's about 51.1 years!), and also that the time can be negative ... something is odd there for me.

Also, I had the VPN lock up when leaving home (WiFi -> 4G network), and this is the tail of the logcat from that time - including the shutdown. BTW: I'm using the PID to capture all messages at the moment.

I'll have a look at the code tonight ... need to get some critical work done first.

[logcat2.txt](https://github.com/AdAway/AdAway/files/5862989/logcat2.txt)

**Comment by tommyjli** [comment]
Installed 5.4.0 with "Monitor connection" but still having problems with VPN connection only when moving from mobile --> Wifi (100% reproduceable); Wifi --> Mobile works fine 100% of the time

**Comment by bobbob234** [comment]
I'm just checking in on the status of this bug. I'm in 5.5.1 and it seems to happen almost every time, if not every time, that I switch from wifi to mobile data.

**Comment by BearTM** [comment]
The same issue is still happening for me. I've become so adept at disconnecting the VPN, closing the app, and restarting it every time I switch networks.

Often the app locks up attempting to update the sources, but it's unable to because the network/DNS is unable to resolve and fails to timeout/restart. It's still some kind of network change issue, that needs to be fixed.

I will capture some logcat files again, and see what I can figure out.

**Comment by empireromain** [comment]
Hello I had this issues until 5 minutes ago...
I tried the way in the first post, but it didn't work for me...

I had to go on the vpn setting to allow ad away

I am on a samsung s20 FE 5g i will post some screenshot bellow.

![Screenshot_20210503-120134_Settings Suggestions](https://user-images.githubusercontent.com/35383194/116864539-4dce1780-ac08-11eb-9e71-ad4d1d710cac.jpg)
![Screenshot_20210503-120140_Settings](https://user-images.githubusercontent.com/35383194/116864549-532b6200-ac08-11eb-9bd1-e613c4481482.jpg)
![Screenshot_20210503-120147_Settings](https://user-images.githubusercontent.com/35383194/116864565-5888ac80-ac08-11eb-8e09-87fbd53e292e.jpg)
![Screenshot_20210503-120155_Settings](https://user-images.githubusercontent.com/35383194/116864572-5c1c3380-ac08-11eb-8326-5f545722a331.jpg)

**Comment by SirLT** [comment]
Does anyone of you have static IP for your devices within the wifi network you're connected to? I'm on holiday for 2 weeks now so I can't verify it right now but I've noticed that this behavior only happens on my own router at home. I never have any problems with wifi -> mobile. Also, I don't seem to lose connection on any other wifi network than my own. The only difference between the wifin networks that I can see is that I have designated IP addresses for my devices on my router. Just now tried connecting to the hotel wifi and didn't have any loss of connection. At home it's 100% failure rate.

**Comment by RobBeyer** [comment]
I'm not sure what happened to this comment from @SirLT:

> Does anyone of you have static IP for your devices within the wifi network you're connected to? I'm on holiday for 2 weeks
> now so I can't verify it right now but I've noticed that this behavior only happens on my own router at home. I never have
> any problems with wifi -> mobile. Also, I don't seem to lose connection on any other wifi network than my own. The only
> difference between the wifin networks that I can see is that I have designated IP addresses for my devices on my router.
> Just now tried connecting to the hotel wifi and didn't have any loss of connection. At home it's 100% failure rate.

I'm also using a static IP, but it is being assigned by DHCP at home. Could the static assignment process have anything to do with the failure(s) to reconnect?

Additionally, while the app is "checking for blacklist updates" the entire UI/interface is blocked, and if during that time the VPN connection has failed, it necessarily waits until a (long) request timeout has passed before it can fail the VPN connection and attempts a reconnect. The only solution is to disconnect and reconnect the VPN. When doing that the App still believes it is connected (stays red, not grey), and closing and re-opening the app solves that display/state bug too.

**Comment by MoL82** [comment]
Tested this, but did not work for me. I have to switch AdAway off and on everytime I switch connections.
Samsung A52s, stock ROM, no root.

> Hello I had this issues until 5 minutes ago... I tried the way in the first post, but it didn't work for me...
> 
> I had to go on the vpn setting to allow ad away
> 
> I am on a samsung s20 FE 5g i will post some screenshot bellow.

**Comment by klodoma** [comment]
Hi All, 

I have the same issue. On Samsung Galaxy S21 non rooted. 

The VPN doesn't work if the network state canges:
- Mobile to WIFI
- No connection
- Activate/Deactivate flightmode

The VPN might be displayed as "Active" but no traffice is let through.

**Comment by hungprocess** [comment]
I think I've found a way to consistently reproduce this behavior, at least on my networks.

Some background: at some point in the past I found that AdAway only seemed to work properly on my cell network when I had IPv6 support enabled in AdAway's settings. Due to this, I've had that setting enabled since then. My cell network supports IPv6, my home router does not. I'm currently running the "AdAway-feat-vpn-220212" build you provided in the last post. 

Steps to reproduce:
1. Have AdAway configured with IPv6 support turned on.
2. Be connected to cell (i.e. IPv6) network.
3. Have AdAway VPN running.
4. Try to pull up a website. Works fine.
5. Switch to home wifi (i.e. IPv4 only) network.
6. Try to pull up a different website. Consistent DNS timeouts on any site I try.
7. Turn AdAway VPN off and back on again.
8. Retry failing sites. All of them work.

I'm including a logcat where I followed the above process. Hopefully I captured it correctly; this is my first time submitting this sort of bug report. I am very much *not* an Android developer, but if I'm reading this log correctly it appears that my DNS requests are still being forwarded to the IPv6 address of my cell network's DNS resolvers even after I've already switched to my home network. Since my home network doesn't route IPv6, those requests consistently time out. After disabling and re-enabling the AdAway VPN, DNS requests are correctly sent to the IPv4 address of my home DNS resolver.

For what it's worth, I tried disabling the IPv6 setting in AdAway just now and following the same process didn't trigger a failure.

[logcat_02-13-2022_00-13-52.txt](https://github.com/AdAway/AdAway/files/8055032/logcat_02-13-2022_00-13-52.txt)


(It may not be relevant to this bug, but I also noticed that using my cell network works fine now even with IPv6 disabled in AdAway. I don't know if the bug that caused me to enable IPv6 has been fixed, or if it was just a coincidence back then that connectivity started working when I enabled it.)

**Comment by PerfectSlayer** [comment]
Hi @hungprocess 

Thanks for the report! I checked the logcat and you are right.
Once connected to WiFi, your DNS requests are still sent to your cellular network, resulting in timeouts.

I might have something that could help your issue.
Can you test this build: [AdAway-feat-vpn-220213.zip](https://github.com/AdAway/AdAway/files/8055727/AdAway-feat-vpn-220213.zip)?

It should always try to restart VPN once a new network is found, not only when there is no more connectivity, and try to fix another crash where a tunnel interface becomes no more a tunnel interface 🤔

```
Failed to initialize connection monitor.
java.lang.IllegalStateException: No successful match so far
	at java.util.regex.Matcher.ensureMatch(Matcher.java:1189)
	at java.util.regex.Matcher.group(Matcher.java:456)
	at org.adaway.vpn.worker.VpnConnectionMonitor.pickLastVpnNetworkInterface(VpnConnectionMonitor.java:75)
	at org.adaway.vpn.worker.VpnConnectionMonitor.findVpnNetworkInterface(VpnConnectionMonitor.java:57)
	at org.adaway.vpn.worker.VpnConnectionMonitor.initialize(VpnConnectionMonitor.java:90)
	at org.adaway.vpn.worker.VpnWorker.runVpn(VpnWorker.java:218)
	at org.adaway.vpn.worker.VpnWorker.work(VpnWorker.java:188)
	at org.adaway.vpn.worker.VpnWorker.$r8$lambda$S7jOMU5kNPTjZYYwMeBRmjMYE5M(Unknown Source:0)
	at org.adaway.vpn.worker.VpnWorker$$ExternalSyntheticLambda1.run(Unknown Source:2)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
```

**Comment by hungprocess** [comment]
Hi @PerfectSlayer, thanks for responding so quickly! I installed the new build and tried the same process again. This time it still fails unfortunately, and even after stopping and restarting AdAway VPN my DNS requests time out. Stopping AdAway VPN entirely restores DNS functionality.

I'm including a logcat. The format of this one is a bit different than the last as I switched to an application that lets me filter by package name. Without it, I was getting ~25MB dumps. If this new format doesn't work for you, please let me know.

[pidcat_02-13_12-23.txt](https://github.com/AdAway/AdAway/files/8056220/pidcat_02-13_12-23.txt)

**Comment by PerfectSlayer** [comment]
Hi, thanks for the test.

This logcat is harder to read as it does no more inclure system events (like ConnectivityManager stuff) nor timings but I can still see some issues.

When you connect to the WiFi, the cellular network is still the default network.
So, when the app is notified of a new network and restart the VPN, it uses the celluar DNS that later times out.
I added a way to restart the VPN when losing a network (cellular) but still having another network available (wifi).
It should make the app use the WiFi DNS servers if you wait a little after joining WiFi (wait for the device to resign cellular connection then current DNS queries timeouts then it *should* work).

Weirdly, at the end of the log, the app keeps using cellular DNS... It could be another issue.
I added more logs to try to understand what happens.

Here is the new build: [AdAway-feat-vpn-220214.zip](https://github.com/AdAway/AdAway/files/8062609/AdAway-feat-vpn-220214.zip)

**Comment by hungprocess** [comment]
Yep, that build seems to work for me, thanks!

**Comment by PerfectSlayer** [comment]
Nice! Glad it finally works 😅 

Could you make a last test with this build? 
I removed the VPN reconnection each time a new network is available. It only reconnect on network lost.
It will be less aggressive and more stable for the future.

If it work (or not), could you still post a logcat (the format you want) so I can have a log of the working behavior too. Thanks!

Here is the new (last?) build: [AdAway-feat-vpn-220215.zip](https://github.com/AdAway/AdAway/files/8066128/AdAway-feat-vpn-220215.zip)

**Comment by hungprocess** [comment]
Apologies for the delay, I was having a hard time consistently reproducing the behavior I mentioned previously. Occasionally, I'd turn AdAway VPN off, the app would turn gray to indicate the VPN was off, but then the VPN key icon in the notification bar would reappear despite the app staying gray. If I had to take a wild guess I'd say I was running into some sort of race condition, but I'm not an expert.

Unfortunately, I think when I reported success last time it was a false positive and my results were a byproduct of this weird "doubling up" of VPNs.


**Test attempt with the 2/14 build:**
1. Connected to cellular network.
2. Started logging.
3. Started AdAway.
4. Loaded a site in Firefox, which succeeded.
5. Switched to wifi network.
6. Attempted to load a site, which failed.
7. Stopped AdAway.
8. VPN key icon disappeared from notification bar, and AdAway app background turned gray.
9. Almost immediately afterward, VPN key icon reappeared in bar. AdAway background still gray.
10. Attempted to load a site, which failed.
11. Turned on AdAway, red background, key icon in bar.
12. Attempted to load site, which failed.
13. Turned off AdAway, gray background, key icon still in bar.
14. Loaded site, which worked.
[Logcat for attempt with 2/14 build](https://github.com/AdAway/AdAway/files/8076012/2-14_build_logcat.txt)


**Test attempt with the 2/15 build:**
1. Connected to cellular network.
2. Started logging.
3. Started AdAway.
4. Loaded a site in Firefox, which succeeded.
5. Switched to wifi network.
6. Attempted to load site, which failed.
7. Attempted to load site again, which failed.
8. Turned off AdAway, gray background, no key in bar.
9. Turned on AdAway, red background, key in bar.
10. Loaded site, which succeeded.
[Logcat for attempt with 2/15 build](https://github.com/AdAway/AdAway/files/8076016/2-15_build_logcat.txt)


**Some observations that aren't reflected in these logs:**
- I encountered the "gray background / key still present" issue with both builds at least once. 
- To keep the logs to a reasonable size, these logs show brief tests. I also tried testing by waiting a minute or more after switching from cellular to wifi to see if the VPN restart code you added would get subsequent queries to work. I can generate longer-running logs for you if you're able to tell me a list of all the logcat tags or packages you're interested in, and I can filter out the extraneous noise.

**Additional information about my phone that might be relevant?:**
- My cellular provider is Google Fi.
- I have two users set up on this phone. To try and minimize the possible interference from the second account, I rebooted the phone before the tests, and didn't log into the second account before testing.
- The phone is a Pixel 3a.

If there's anything else I can do to help track down what's going on, please let me know. Thanks again for all your great work!

**Comment by PerfectSlayer** [comment]
Thanks again for the very complete report!

I mainly focused on the 2/15 one as the 2/14 will leads to race condition and stability issue (it will try to restart the VPN too often).

About the 2/15 build, I find an issue I totally missed: when you connect to the WiFi network, the VPN is restarted but the DNS used are still the ones from your cellular connection... 

I already found an issue with this part of code (using the VPN fake DNS, used to intercept DNS traffic, instead of the underlying network (WiFi or cellular) DNS). So I ignore the active network to get DNS (as they are fake) and pick another (random) one that is not tied to a VPN. In your case, your cellular VPN whereas you just connected to WiFi.

So I improve the logic used to find the true DNS servers:
- If no active network, use any non vpn network DNS,
- If active network is defined:
  - If active network is not tied to a VPN, use its DNS
  - If active network is tied to a VPN, find any network that matches the transport type (cellular or WiFi) that is not tied to a VPN and use its DNS.

This issue does (using the wrong network DNS) was not relevant for me as both network DNS (cellular and network) are accessible from any network (both 5g and wifi comes from the same provider). With your help, I could finally find and test this kind of trouble 🙏 

Here is the new build: [AdAway-feat-vpn-229216.zip](https://github.com/AdAway/AdAway/files/8081049/AdAway-feat-vpn-229216.zip)

Keep in mind this build is based on 2/15 so it will wait for Android to notify the app that the WiFi network is validated then the cellular network is no more available to restart the VPN. It could take about ~20s but should not be noticeable (as the cellular network is still here until the handover). So it could be useful to wait for ~30s after WiFi connection to test your WiFi network using Firefox

**Comment by hungprocess** [comment]
It works like a charm! 😃

As you pointed out, there's a short period during the handover when DNS fails, but even then it works again by the time I can reload the page. Here's a logcat of me switching around between cellular and two different wifi networks. Great job!

[logcat_02-16-2022_21-34-24.txt](https://github.com/AdAway/AdAway/files/8085125/logcat_02-16-2022_21-34-24.txt)

**Comment by PerfectSlayer** [comment]
Awesome!

If you can, keep using it for few days. You still might encounter new issues 😅 
On my side, I will have a look to find how I can improve the handover (like recreating the DNS request on the new network before it times out).

Thanks again for all your tests!

**Comment by drothenberger** [comment]
I have noticed a problem with the latest AdAway-feat-vpn-229216 build. The VPN enables itself even though I have manually turned off AdAway. Sometimes it happens quickly, and sometimes it takes a while, but it always happens. The VPN icon appears in the notification bar, but the app looks as if it is off.

I collected a logcat for when this happened. It's pretty big, though, since I don't know how to force this to happen.

[Log_2022-02-18_11-14-31.zip](https://github.com/AdAway/AdAway/files/8099966/Log_2022-02-18_11-14-31.zip)

**Comment by PerfectSlayer** [comment]
Hi @drothenberger 

Thanks for the report! I can see from your log the VPN is restarted due the the VPN service hearbeat.
I found an issue where the heartbeat was not stopped when stopping the VPN service and I fixed it so it's a good start.
There might be another issue. I will keep investigating and make a new build soon(™️)

**Comment by MoL82** [comment]
> Hi @drothenberger
> 
> Thanks for the report! I can see from your log the VPN is restarted due the the VPN service hearbeat. I found an issue where the heartbeat was not stopped when stopping the VPN service and [I fixed it](c707bf6a9a21662bf094bce55575779a12a4057b) so it's a good start. There might be another issue. I will keep investigating and make a new build soon(™️)
> 
> EDIT: Yes, I found another issue. The heartbeat was only testing if the VPN service is running but did not take into account if the user stopped it. [I fixed it](e5af18eec6a0428cb7b647d3652ccfb47bcb67a3).
> 
> Here is the new build: [AdAway-feat-vpn-220220.zip](https://github.com/AdAway/AdAway/files/8103661/AdAway-feat-vpn-220220.zip) As usual, we keep trying it and report if something looks weird. Thanks 🙏

This build works very well. Thank you very much for your effort.

**Comment by PerfectSlayer** [comment]
You can finally stop using dev build and switch to the official beta channel! 🎉 
[The new 6.x versions](https://github.com/AdAway/AdAway/releases/tag/v6.0.0b) contains all the fixed you need :)
